<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>seqgroup tour • seqgroup</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="seqgroup tour">
<meta property="og:description" content="seqgroup">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqgroup</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/seqgroup_tour.html">seqgroup tour</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>seqgroup tour</h1>
                        <h4 class="author">Karoline Faust</h4>
            
            <h4 class="date">2020-04-30</h4>
      
      
      <div class="hidden name"><code>seqgroup_tour.Rmd</code></div>

    </div>

    
    
<div id="background-on-the-example-data" class="section level2">
<h2 class="hasAnchor">
<a href="#background-on-the-example-data" class="anchor"></a>Background on the example data</h2>
<p>The example data in this tutorial are taken from an IBD study carried out by Franzosa et al. (<a href="https://www.nature.com/articles/s41564-018-0306-4" class="uri">https://www.nature.com/articles/s41564-018-0306-4</a>). In this study, faecal samples from the PRISM cohort were sequenced and analysed. The PRISM cohort consists of 68 cases of Crohn’s disease (CD), 53 cases of ulcerative colitis (UC) and 34 controls, thus totalling 155 samples. Whole-genome shotgun sequencing was carried out. Relative abundances of species were estimated with species-specific marker genes using MetaPhlan2 (<a href="http://huttenhower.sph.harvard.edu/metaphlan2" class="uri">http://huttenhower.sph.harvard.edu/metaphlan2</a>) and function abundances were computed with HUMAnN2. Species below or equal to 0.1% abundance in at least five samples were excluded. The abundance of functions (i.e. enzyme-coding genes) was quantified using the HUManN2 pipeline (<a href="http://huttenhower.sph.harvard.edu/humann2" class="uri">http://huttenhower.sph.harvard.edu/humann2</a>). CD and UC together are also referred to as inflammatory bowel disease (IBD).</p>
</div>
<div id="preparation-of-the-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#preparation-of-the-analysis" class="anchor"></a>Preparation of the analysis</h2>
<p>We start by loading the seqgroup library.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">seqgroup</span>)</pre></body></html></div>
<p>Next, we load the taxon abundances, taxon lineages, functions and sample metadata from the PRISM cohort. Taxon abundances and metadata have matching sample identifiers, thus the status of a sample can be looked up in the metadata table. We summarise the metadata using the summary command. This command also tells us that some metadata information are missing, for instance there are 62 missing values (NA) for fecal calprotectin.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"ibd_taxa"</span>) <span class="co"># taxon x sample matrix</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"ibd_lineages"</span>) <span class="co"># lineage x taxon matrix</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"ibd_metadata"</span>) <span class="co"># metadata x sample data frame</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"ibd_functions"</span>) <span class="co"># functions x sample matrix</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">ibd_metadata</span>)</pre></body></html></div>
<pre><code>##  SRA_metagenome_name      Age        Diagnosis  Fecal.Calprotectin antibiotic
##  G35127 :  1         24     : 10   CD     :68   2      : 5         No  :134  
##  G35128 :  1         30     : 10   Control:34   231    : 3         Yes : 18  
##  G35152 :  1         26     :  7   UC     :53   32     : 3         NA's:  3  
##  G36347 :  1         29     :  7                1      : 2                   
##  G36348 :  1         31     :  7                12     : 2                   
##  G36349 :  1         32     :  6                (Other):78                   
##  (Other):149         (Other):108                NA's   :62                   
##  immunosuppressant mesalamine steroids  
##  No :100           No  :109   No  :122  
##  Yes: 55           Yes : 44   Yes : 31  
##                    NA's:  2   NA's:  2  
##                                         
##                                         
##                                         
## </code></pre>
<p>In this data set, we have three groups of interest: CD, UC and control. For future analyses, it is handy to have a vector with the group memberships ready. We can extract this vector from the metadata table. The two disease groups (UCD and CD) can also be merged into a single group. This is done in the second group membership vector (ibd.groups).</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">groups</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span>(<span class="no">ibd_metadata</span>$<span class="no">Diagnosis</span>)
<span class="no">ibd.groups</span><span class="kw">=</span><span class="no">groups</span>
<span class="no">ibd.groups</span>[<span class="no">ibd.groups</span><span class="kw">==</span><span class="st">"UC"</span>]<span class="kw">=</span><span class="st">"IBD"</span>
<span class="no">ibd.groups</span>[<span class="no">ibd.groups</span><span class="kw">==</span><span class="st">"CD"</span>]<span class="kw">=</span><span class="st">"IBD"</span></pre></body></html></div>
</div>
<div id="exploration-of-diversity" class="section level2">
<h2 class="hasAnchor">
<a href="#exploration-of-diversity" class="anchor"></a>Exploration of diversity</h2>
<p>The first question when exploring a data set of microbial abundances is often whether groups have significantly different alpha or beta diversity. Seqgroup offers the compareGroups function to compare alpha and beta diversity across groups. Alpha diversity is assessed with the Shannon index, whereas beta diversity can be assessed for instance by comparing the Bray Curtis dissimilarity distribution. The number of samples considered for each group is by default equalised by randomly subsampling groups to the smallest group. Here, the alpha diversity of the control group is significantly higher while its beta diversity is significantly lower compared to the two IBD groups according to the unpaired Wilcoxon test. Interestingly, not only the taxonomic but also the function beta diversity is higher in the IBD groups than in the control samples. Function stat_compare_means from the R package ggpubr is used to plot the p-value on top of the distributions.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="../reference/compareGroups.html">compareGroups</a></span>(<span class="no">ibd_taxa</span>,<span class="kw">groups</span><span class="kw">=</span><span class="no">groups</span>,<span class="kw">property</span><span class="kw">=</span><span class="st">"alpha"</span>,<span class="kw">pvalViz</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<pre><code>## [1] "Constraining sample number randomly to the same minimal group sample number of 34"
## [1] "Processing group CD"
## [1] "Processing group UC"
## [1] "Processing group Control"
## [1] "Number of significant differences in property alpha across groups: 2"</code></pre>
<p><img src="figure_seqgroup_tour/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="../reference/compareGroups.html">compareGroups</a></span>(<span class="no">ibd_taxa</span>,<span class="kw">groups</span><span class="kw">=</span><span class="no">groups</span>,<span class="kw">property</span><span class="kw">=</span><span class="st">"beta"</span>,<span class="kw">pvalViz</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<pre><code>## [1] "Constraining sample number randomly to the same minimal group sample number of 34"
## [1] "Processing group CD"
## [1] "Processing group UC"
## [1] "Processing group Control"
## [1] "Number of significant differences in property beta across groups: 3"</code></pre>
<p><img src="figure_seqgroup_tour/unnamed-chunk-4-2.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="../reference/compareGroups.html">compareGroups</a></span>(<span class="no">ibd_functions</span>,<span class="kw">groups</span><span class="kw">=</span><span class="no">groups</span>,<span class="kw">property</span><span class="kw">=</span><span class="st">"beta"</span>,<span class="kw">pvalViz</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<pre><code>## [1] "Constraining sample number randomly to the same minimal group sample number of 34"
## [1] "Processing group CD"
## [1] "Processing group UC"
## [1] "Processing group Control"
## [1] "Number of significant differences in property beta across groups: 3"</code></pre>
<p><img src="figure_seqgroup_tour/unnamed-chunk-4-3.png" width="576" style="display: block; margin: auto;"></p>
<p>The assessment of significance of a difference in beta diversity using the Bray Curtis distribution is problematic, since the number of all possible sample pairs is so large that even small differences become significant. Alternatively, we can assess beta diversity by fitting the Dirichlet-Multinomial to group-specific sample sets and compare the overdispersion parameter theta. In this case, the overdisperson supports the results obtained by comparing the Bray-Curtis distribution.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="../reference/compareGroups.html">compareGroups</a></span>(<span class="no">ibd_taxa</span>,<span class="kw">groups</span><span class="kw">=</span><span class="no">groups</span>,<span class="kw">property</span><span class="kw">=</span><span class="st">"beta"</span>,<span class="kw">method</span><span class="kw">=</span><span class="st">"DM"</span>)</pre></body></html></div>
<pre><code>## [1] "Abundances have been scaled with factor: 10000"
## [1] "Constraining sample number randomly to the same minimal group sample number of 34"
## [1] "Processing group CD"
## [1] "Processing group UC"
## [1] "Processing group Control"
## NULL</code></pre>
<p><img src="figure_seqgroup_tour/unnamed-chunk-5-1.png" width="288" style="display: block; margin: auto;"></p>
</div>
<div id="taxon-composition" class="section level2">
<h2 class="hasAnchor">
<a href="#taxon-composition" class="anchor"></a>Taxon composition</h2>
<p>Taxon composition is conveniently visualised by sorting taxa by abundance, for instance by their sum across samples, and then displaying only the top 10 or 15 taxa. We can do this for a single group. In the bar plot below, we can see that despite representing relative abundances, samples do not sum to one, because of the filtering step mentioned above.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">control.indices</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">groups</span><span class="kw">==</span><span class="st">"Control"</span>)
<span class="fu"><a href="../reference/groupBarplot.html">groupBarplot</a></span>(<span class="no">ibd_taxa</span>[,<span class="no">control.indices</span>],<span class="kw">topTaxa</span> <span class="kw">=</span> <span class="fl">15</span>,<span class="kw">legend.shift</span> <span class="kw">=</span> <span class="fl">1.5</span>, <span class="kw">main</span><span class="kw">=</span><span class="st">"Control"</span>)</pre></body></html></div>
<p><img src="figure_seqgroup_tour/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
<p>We can display the same randomly selected number of samples for the three groups. However, the taxa are now ranked by their sum across all samples, instead of being ranked per group.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="../reference/groupBarplot.html">groupBarplot</a></span>(<span class="no">ibd_taxa</span>,<span class="kw">topTaxa</span> <span class="kw">=</span> <span class="fl">15</span>,<span class="kw">groups</span><span class="kw">=</span><span class="no">groups</span>, <span class="kw">randSampleNum</span> <span class="kw">=</span> <span class="fl">10</span>, <span class="kw">legend.shift</span> <span class="kw">=</span> <span class="fl">1.5</span>, <span class="kw">main</span><span class="kw">=</span><span class="st">"Taxon composition"</span>)</pre></body></html></div>
<pre><code>## [1] "Number of samples in group CD : 68"
## [1] "Number of samples in group UC : 53"
## [1] "Number of samples in group Control : 34"</code></pre>
<p><img src="figure_seqgroup_tour/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;"></p>
<p>We can obtain a group-specific ranking and preserve the color code using extendTaxonColorMap. This illustrates that IBD samples have different top-abundant taxa than control samples do.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">ibd.indices</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">ibd.groups</span><span class="kw">==</span><span class="st">"IBD"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>,<span class="fl">1</span>))
<span class="no">taxon.color.map</span><span class="kw">=</span><span class="fu"><a href="../reference/groupBarplot.html">groupBarplot</a></span>(<span class="no">ibd_taxa</span>[,<span class="no">control.indices</span>],<span class="kw">topTaxa</span> <span class="kw">=</span> <span class="fl">7</span>, <span class="kw">randSampleNum</span> <span class="kw">=</span> <span class="fl">20</span>, <span class="kw">extendTaxonColorMap</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">main</span><span class="kw">=</span><span class="st">"Control"</span>)</pre></body></html></div>
<pre><code>## [1] "Number of samples in group all : 34"</code></pre>
<div class="sourceCode"><html><body><pre class="r"><span class="no">taxon.color.map</span><span class="kw">=</span><span class="fu"><a href="../reference/groupBarplot.html">groupBarplot</a></span>(<span class="no">ibd_taxa</span>[,<span class="no">ibd.indices</span>],<span class="kw">topTaxa</span> <span class="kw">=</span> <span class="fl">7</span>, <span class="kw">randSampleNum</span> <span class="kw">=</span> <span class="fl">20</span>, <span class="kw">taxon.color.map</span> <span class="kw">=</span> <span class="no">taxon.color.map</span>, <span class="kw">extendTaxonColorMap</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">main</span><span class="kw">=</span><span class="st">"IBD"</span>)</pre></body></html></div>
<pre><code>## [1] "Number of samples in group all : 121"</code></pre>
<p><img src="figure_seqgroup_tour/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">1</span>))</pre></body></html></div>
<p>We can also obtain a complementary view on the taxon composition with phyloseq, which is supported by seqgroup.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">phyloseq</span>)
<span class="no">prism.obj</span><span class="kw">=</span><span class="fu"><a href="../reference/toPhyloseq.html">toPhyloseq</a></span>(<span class="no">ibd_taxa</span>,<span class="kw">metadata</span><span class="kw">=</span><span class="no">ibd_metadata</span>,<span class="kw">lineages</span><span class="kw">=</span><span class="no">ibd_lineages</span>)
<span class="no">p</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/plot_bar.html">plot_bar</a></span>(<span class="no">prism.obj</span>, <span class="st">"Order"</span>, <span class="kw">fill</span><span class="kw">=</span><span class="st">"Order"</span>, <span class="kw">facet_grid</span><span class="kw">=</span>~<span class="no">Diagnosis</span>) + <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="kw">color</span><span class="kw">=</span><span class="no">Order</span>, <span class="kw">fill</span><span class="kw">=</span><span class="no">Order</span>), <span class="kw">stat</span><span class="kw">=</span><span class="st">"identity"</span>, <span class="kw">position</span><span class="kw">=</span><span class="st">"stack"</span>)
<span class="no">p</span></pre></body></html></div>
<p><img src="figure_seqgroup_tour/unnamed-chunk-9-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="ordination" class="section level2">
<h2 class="hasAnchor">
<a href="#ordination" class="anchor"></a>Ordination</h2>
<p>To check whether samples group by community composition, an ordination is frequently carried out. In seqgroup, vegan’s ordination functions are used to do bi- and triplots. By default, Bray Curtis is used to assess sample dissimilarities.</p>
<p>A biplot is a plot that shows both samples and taxa in the same ordination. The function seqPCoA identifies taxa that co-vary significantly with the principal components (that is the eigen vectors of the dissimilarity matrix), which represent the main axes of community variaton. Significance of covariation is assessed through a permutation test for the selected number of top covarying taxa. The resulting plot shows that control samples are grouped separately from most IBD samples as seen in Figure 1c in Franzosa et al. (<a href="https://www.nature.com/articles/s41564-018-0306-4" class="uri">https://www.nature.com/articles/s41564-018-0306-4</a>) and that an unclassified Escherichia points in the direction of the IBD sample cluster. This means that Escherichia abundance is higher in samples further away from the center along the direction of its arrow. Bray-Curtis is robust to co-absences, which bias other measures of dissimilarity/distance. For instance, when applying a prevalence filter and repeating the plot, the sample groups are preserved.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="../reference/seqPCoA.html">seqPCoA</a></span>(<span class="no">ibd_taxa</span>,<span class="kw">groups</span><span class="kw">=</span><span class="no">groups</span>, <span class="kw">topTaxa</span><span class="kw">=</span><span class="fl">30</span>)</pre></body></html></div>
<p><img src="figure_seqgroup_tour/unnamed-chunk-10-1.png" width="576" style="display: block; margin: auto;"></p>
<pre><code>## [1] "Cluster quality index silhouette"
## [1] -0.02240243
## [1] "Among the top 30 covarying taxa, 2 are significant."
## [1] "Subdoligranulum_unclassified"
## [1] "Escherichia_unclassified"</code></pre>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># prevalence filter</span>
<span class="no">prev</span><span class="kw">=</span><span class="fl">50</span> <span class="co"># minimum occurrence in 50 samples</span>
<span class="no">ibd_taxa_incidence</span><span class="kw">=</span><span class="no">ibd_taxa</span>
<span class="no">ibd_taxa_incidence</span>[<span class="no">ibd_taxa_incidence</span><span class="kw">&gt;</span><span class="fl">0</span>]<span class="kw">=</span><span class="fl">1</span>
<span class="no">rowSums</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">ibd_taxa_incidence</span>)
<span class="no">indices.prev</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">rowSums</span><span class="kw">&gt;=</span><span class="no">prev</span>)
<span class="fu"><a href="../reference/seqPCoA.html">seqPCoA</a></span>(<span class="no">ibd_taxa</span>[<span class="no">indices.prev</span>,], <span class="kw">groups</span><span class="kw">=</span><span class="no">groups</span>, <span class="kw">topTaxa</span><span class="kw">=</span><span class="fl">30</span>)</pre></body></html></div>
<p><img src="figure_seqgroup_tour/unnamed-chunk-10-2.png" width="576" style="display: block; margin: auto;"></p>
<pre><code>## [1] "Cluster quality index silhouette"
## [1] -0.02462392
## [1] "Among the top 30 covarying taxa, 4 are significant."
## [1] "Prevotella_copri"
## [1] "Dorea_longicatena"
## [1] "Eubacterium_hallii"
## [1] "Enterococcus_faecium"</code></pre>
<p>In a triplot, metadata covarying with the principal components are plotted in addition. Seqgroup wraps vegan’s envfit function to do so, which does not accept missing values. In addition, envfit needs to know whether to treat metadata as numeric or categoric. By default, they are all categoric, even age, which is better treated as numeric here. Thus, before carrying out a triplot with seqPCoA, we need to assign the correct data categories and remove metadata and/or samples with missing values. The ordination then shows that fecal calprotectin values covary strongly and significantly with the community composition of IBD samples, in agreement with the results of Franzosa and colleagues.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># assign numeric metadata types, metadata with only 2 values are treated automatically as categoric (binary).</span>
<span class="no">ibd_metadata</span><span class="kw">=</span><span class="fu"><a href="../reference/assignMetadataTypes.html">assignMetadataTypes</a></span>(<span class="no">ibd_metadata</span>,<span class="kw">categoric</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"SRA_metagenome_name"</span>,<span class="st">"Diagnosis"</span>))</pre></body></html></div>
<pre><code>## [1] "Metadata SRA_metagenome_name is categoric"
## [1] "Metadata Age is numeric"
## [1] "Metadata Diagnosis is categoric"
## [1] "Metadata Fecal.Calprotectin is numeric"
## [1] "Metadata antibiotic is binary"
## [1] "Metadata immunosuppressant is binary"
## [1] "Metadata mesalamine is binary"
## [1] "Metadata steroids is binary"</code></pre>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># remove 65 samples with missing calprotectin measurements or other missing values in the metadata</span>
<span class="no">na.indices</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">ibd_metadata</span>),<span class="kw">arr.ind</span><span class="kw">=</span><span class="fl">TRUE</span>)[,<span class="fl">1</span>])
<span class="no">indices.to.keep</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html">setdiff</a></span>(<span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">ibd_metadata</span>),<span class="no">na.indices</span>)
<span class="no">ibd_metadata.nafree</span><span class="kw">=</span><span class="no">ibd_metadata</span>[<span class="no">indices.to.keep</span>,]
<span class="no">ibd_taxa.nafree</span><span class="kw">=</span><span class="no">ibd_taxa</span>[,<span class="no">indices.to.keep</span>]
<span class="fu"><a href="../reference/seqPCoA.html">seqPCoA</a></span>(<span class="no">ibd_taxa.nafree</span>,<span class="kw">metadata</span><span class="kw">=</span><span class="no">ibd_metadata.nafree</span>,<span class="kw">groups</span><span class="kw">=</span><span class="no">groups</span>[<span class="no">indices.to.keep</span>],<span class="kw">topTaxa</span><span class="kw">=</span><span class="fl">60</span>)</pre></body></html></div>
<pre><code>## [1] "1 significant numeric metadata found, in order of significance:"
## [1] "Fecal.Calprotectin"
## [1] "4 significant categoric metadata found, in order of significance:"
## [1] "Diagnosis"
## [1] "antibiotic"
## [1] "immunosuppressant"
## [1] "steroids"</code></pre>
<p><img src="figure_seqgroup_tour/unnamed-chunk-11-1.png" width="576" style="display: block; margin: auto;"></p>
<pre><code>## [1] "Cluster quality index silhouette"
## [1] -0.01835481
## [1] "Among the top 60 covarying taxa, 6 are significant."
## [1] "Faecalibacterium_prausnitzii"
## [1] "Ruminococcus_bromii"
## [1] "Streptococcus_thermophilus"
## [1] "Ruminococcus_obeum"
## [1] "Dorea_formicigenerans"
## [1] "Megasphaera_micronuciformis"</code></pre>
<p>The main results (IBD and control samples cluster apart, calprotectin is increased in IBD samples) are robust to the dissimilarity/distance chosen to compute the ordination and are also found with the Gower distance.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="../reference/seqPCoA.html">seqPCoA</a></span>(<span class="no">ibd_taxa.nafree</span>,<span class="kw">metadata</span><span class="kw">=</span><span class="no">ibd_metadata.nafree</span>,<span class="kw">groups</span><span class="kw">=</span><span class="no">groups</span>[<span class="no">indices.to.keep</span>],<span class="kw">topTaxa</span><span class="kw">=</span><span class="fl">60</span>, <span class="kw">dis</span><span class="kw">=</span><span class="st">"gower"</span>,<span class="kw">xlim</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(-<span class="fl">0.3</span>,<span class="fl">0.3</span>),<span class="kw">main</span><span class="kw">=</span><span class="st">"Gower"</span>)</pre></body></html></div>
<pre><code>## [1] "1 significant numeric metadata found, in order of significance:"
## [1] "Fecal.Calprotectin"
## [1] "4 significant categoric metadata found, in order of significance:"
## [1] "Diagnosis"
## [1] "antibiotic"
## [1] "immunosuppressant"
## [1] "steroids"</code></pre>
<p><img src="figure_seqgroup_tour/unnamed-chunk-12-1.png" width="576" style="display: block; margin: auto;"></p>
<pre><code>## [1] "Cluster quality index silhouette"
## [1] 0.01482298
## [1] "Among the top 60 covarying taxa, 3 are significant."
## [1] "Coprococcus_sp_ART55_1"
## [1] "Coprobacillus_unclassified"
## [1] "Oscillibacter_unclassified"</code></pre>
<p>Faecalibacterium prausnitzii was found to significantly covary with community composition in the reduced missing-value free data set. We can compare its distribution in IBD and control samples. In this case, there is indeed a significant difference: the abundance of F. prausnitzii is reduced in IBD samples (high density at low abundance). Here, only a single taxon is compared across two groups, but a systematic comparison of taxon abundances across groups would require a multiple testing correction. ALDEx2 (<a href="https://www.bioconductor.org/packages/release/bioc/html/ALDEx2.html" class="uri">https://www.bioconductor.org/packages/release/bioc/html/ALDEx2.html</a>) is a good tool for this task.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">ibd.groups.nafree</span><span class="kw">=</span><span class="no">ibd.groups</span>[<span class="no">indices.to.keep</span>]
<span class="no">taxon</span><span class="kw">=</span><span class="st">"Faecalibacterium_prausnitzii"</span>
<span class="fu"><a href="../reference/compareDistribs.html">compareDistribs</a></span>(<span class="no">ibd_taxa.nafree</span>,<span class="kw">taxon</span><span class="kw">=</span><span class="no">taxon</span>,<span class="kw">groups</span><span class="kw">=</span><span class="no">ibd.groups.nafree</span>,<span class="kw">group1</span><span class="kw">=</span><span class="st">"IBD"</span>,<span class="kw">group2</span><span class="kw">=</span><span class="st">"Control"</span>)</pre></body></html></div>
<p><img src="figure_seqgroup_tour/unnamed-chunk-13-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="sample-wise-clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-wise-clustering" class="anchor"></a>Sample-wise clustering</h2>
<p>We can check if unsupervised clustering uncovers the three groups we know. DMM clustering implemented in the DirichletMultinomial package is often used for sample-wise clustering of microbial abundances, though many alternatives exist. Here, we will use the Calinski Harabasz index implemented in the clusterCrit package to determine an optimal cluster number for DMM clustering. For this tutorial, we will test the quality of up to 5 clusters. The agreement of clusters with groups can then be inspected visually in an ordination.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">clusters</span><span class="kw">=</span><span class="fu"><a href="../reference/findClusters.html">findClusters</a></span>(<span class="no">ibd_taxa</span>,<span class="kw">k</span><span class="kw">=</span><span class="fl">NA</span>,<span class="kw">maxK</span><span class="kw">=</span><span class="fl">5</span>,<span class="kw">qualityIndex</span> <span class="kw">=</span> <span class="st">"CH"</span>)
<span class="fu"><a href="../reference/seqPCoA.html">seqPCoA</a></span>(<span class="no">ibd_taxa</span>,<span class="kw">groups</span><span class="kw">=</span><span class="no">groups</span>,<span class="kw">clusters</span><span class="kw">=</span><span class="no">clusters</span>, <span class="kw">topTaxa</span><span class="kw">=</span><span class="fl">30</span>)</pre></body></html></div>
<p><img src="figure_seqgroup_tour/unnamed-chunk-14-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The significance of the agreement between clusters and groups can be assessed for instance with a Chi^2 test.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">clus.table</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">ibd.groups</span>, <span class="no">clusters</span>)
<span class="fu"><a href="https://rdrr.io/r/stats/chisq.test.html">chisq.test</a></span>(<span class="no">clus.table</span>)</pre></body></html></div>
<pre><code>## 
##  Pearson's Chi-squared test with Yates' continuity correction
## 
## data:  clus.table
## X-squared = 25.457, df = 1, p-value = 4.523e-07</code></pre>
</div>
<div id="network-inference" class="section level2">
<h2 class="hasAnchor">
<a href="#network-inference" class="anchor"></a>Network inference</h2>
<p>A network is a straightforward way to visualise associations between taxa. In a network, a node represents a taxon and an edge connecting two taxa represents a significant or robust positive or negative association between their two abundance profiles.</p>
<p>For the purpose of network construction, seqgroup offers the buildNetwork function, which wraps several network inference algorithms, namely those offered in SpiecEasi (<a href="https://github.com/zdk123/SpiecEasi" class="uri">https://github.com/zdk123/SpiecEasi</a>) and in bnlearn (<a href="http://www.bnlearn.com/" class="uri">http://www.bnlearn.com/</a>) as well as a barebones R implementation of CoNet (<a href="http://systemsbiology.vub.ac.be/conet" class="uri">http://systemsbiology.vub.ac.be/conet</a>).</p>
<p>CoNet allows combining several association measures. We will make use of this below, where we will construct networks without computing p-values (which takes too long for a tutorial), but keeping only the edges supported by both selected methods. Note that only those taxa with a prevalence equal to or above the prevalence threshold applied previously are considered.</p>
<p>The resulting networks for IBD and control samples differ, as expected based on the differences in taxon composition. For instance, the IBD network has a larger proportion of nodes belonging to class Clostridia. The color code shows that there is some phylogenetic assortativity, that is taxa are often associated to other taxa of the same class. However, there is no striking difference in network properties.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">def.par</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">no.readonly</span> <span class="kw">=</span> <span class="fl">TRUE</span>) <span class="co"># save previous par settings</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/layout.html">layout</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">2</span>,<span class="fl">4</span>),<span class="fl">2</span>,<span class="fl">2</span>,<span class="kw">byrow</span> <span class="kw">=</span> <span class="fl">TRUE</span>), <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">2</span>,<span class="fl">3</span>), <span class="fl">TRUE</span>) <span class="co"># define space for 4 figures</span>
<span class="no">ibd.network</span><span class="kw">=</span><span class="fu"><a href="../reference/buildNetwork.html">buildNetwork</a></span>(<span class="no">ibd_taxa</span>[<span class="no">indices.prev</span>,<span class="no">ibd.indices</span>], <span class="kw">lineages</span><span class="kw">=</span><span class="no">ibd_lineages</span>[<span class="no">indices.prev</span>,], <span class="kw">method</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"spearman"</span>,<span class="st">"bray"</span>), <span class="kw">repNum</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">nameLevel</span><span class="kw">=</span><span class="st">"species"</span>) <span class="co"># run CoNet on IBD samples</span></pre></body></html></div>
<pre><code>## [1] "Network construction with method(s): spearman, bray"
## [1] "Upper and lower threshold are ignored when more than one method is selected."
## [1] "Processing method spearman"
## [1] "Lower threshold for initial edge number (100): -0.281494202923687"
## [1] "Upper threshold for initial edge number (100): 0.50506016053175"
## [1] "Renormalisation is FALSE"
## [1] "P-values of correlations are computed with cor.test FALSE"
## [1] "Permutations and bootstraps are both computed FALSE"
## [1] "Processing method bray"
## [1] "Lower threshold for initial edge number (100): 0.704900943414939"
## [1] "Upper threshold for initial edge number (100): 0.996359149597926"
## [1] "Renormalisation is FALSE"
## [1] "P-values of correlations are computed with cor.test FALSE"
## [1] "Permutations and bootstraps are both computed FALSE"
## [1] "Associations computed for 109 taxa."
## [1] "Merging methods..."
## [1] "Network has 27 edges."</code></pre>
<div class="sourceCode"><html><body><pre class="r"><span class="no">control.network</span><span class="kw">=</span><span class="fu"><a href="../reference/buildNetwork.html">buildNetwork</a></span>(<span class="no">ibd_taxa</span>[<span class="no">indices.prev</span>,<span class="no">control.indices</span>], <span class="kw">lineages</span><span class="kw">=</span><span class="no">ibd_lineages</span>[<span class="no">indices.prev</span>,], <span class="kw">method</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"spearman"</span>,<span class="st">"bray"</span>), <span class="kw">repNum</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">nameLevel</span><span class="kw">=</span><span class="st">"species"</span>) <span class="co"># run CoNet on control samples</span></pre></body></html></div>
<pre><code>## [1] "Network construction with method(s): spearman, bray"
## [1] "Upper and lower threshold are ignored when more than one method is selected."
## [1] "Processing method spearman"
## [1] "Lower threshold for initial edge number (100): -0.392500839940482"
## [1] "Upper threshold for initial edge number (100): 0.510574581966211"
## [1] "Renormalisation is FALSE"
## [1] "P-values of correlations are computed with cor.test FALSE"
## [1] "Permutations and bootstraps are both computed FALSE"
## [1] "Processing method bray"
## [1] "Lower threshold for initial edge number (100): 0.543878403265414"
## [1] "Upper threshold for initial edge number (100): 0.9995026855208"
## [1] "Renormalisation is FALSE"
## [1] "P-values of correlations are computed with cor.test FALSE"
## [1] "Permutations and bootstraps are both computed FALSE"
## [1] "Associations computed for 109 taxa."
## [1] "Merging methods..."
## [1] "Network has 19 edges."</code></pre>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">ibd.network</span>,<span class="kw">main</span><span class="kw">=</span><span class="st">"IBD network"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">control.network</span>, <span class="kw">main</span><span class="kw">=</span><span class="st">"Control network"</span>)</pre></body></html></div>
<p><img src="figure_seqgroup_tour/unnamed-chunk-16-1.png" width="960" style="display: block; margin: auto;"></p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="no">def.par</span>)</pre></body></html></div>
<p>In general, the results of network inference are strongly dependent on the choice of the inference algorithm. To demonstrate this, we now compute the control network with ARACNE (<a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-7-S1-S7" class="uri">https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-7-S1-S7</a>) as implemented in bnlearn. Very roughly, ARACNE computes all pair-wise mutual information edges and then removes edges with lower mutual information than neighboring edges in order to reduce the number of indirect edges. Here, we only retain edges that are stable across 100 bootstrap iterations, that is repetitions of ARACNE on sub-samples of the data. Since mutual information is a general measure of dependency, no negative or positive edge sign can be assigned. In this case, the number of inferred edges is so large that the result is a typical “hairball”, which requires further analysis. Both igraph and Cytoscape (<a href="https://cytoscape.org/" class="uri">https://cytoscape.org/</a>) offer a range of functions/plugins for network analysis. If Cytoscape is open and the RCy3 package (<a href="https://bioconductor.org/packages/release/bioc/html/RCy3.html" class="uri">https://bioconductor.org/packages/release/bioc/html/RCy3.html</a>) is installed, the network can be sent to Cytoscape directly using RCy3’s function createNetworkFromIgraph().</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">control.aracne</span><span class="kw">=</span><span class="fu"><a href="../reference/buildNetwork.html">buildNetwork</a></span>(<span class="no">ibd_taxa</span>[<span class="no">indices.prev</span>,<span class="no">control.indices</span>],<span class="kw">lineages</span><span class="kw">=</span><span class="no">ibd_lineages</span>[<span class="no">indices.prev</span>,],<span class="kw">method</span><span class="kw">=</span><span class="st">"aracne"</span>,<span class="kw">repNum</span><span class="kw">=</span><span class="fl">100</span>)</pre></body></html></div>
<pre><code>## 
##   Bayesian network learned via Pairwise Mutual Information methods
## 
##   model:
##     [undirected graph]
##   nodes:                                 109 
##   arcs:                                  362 
##     undirected arcs:                     362 
##     directed arcs:                       0 
##   average markov blanket size:           6.64 
##   average neighbourhood size:            6.64 
##   average branching factor:              0.00 
## 
##   learning algorithm:                    ARACNE 
##   mutual information estimator:          Maximum Likelihood (Gauss.) 
##   tests used in the learning procedure:  5886 
## 
## [1] "Carrying out  100 bootstraps to assess edge strengths..."</code></pre>
<p><img src="figure_seqgroup_tour/unnamed-chunk-17-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">control.aracne</span>, <span class="kw">main</span><span class="kw">=</span><span class="st">"Control network inferred with ARACNE"</span>)</pre></body></html></div>
<p><img src="figure_seqgroup_tour/unnamed-chunk-17-2.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode"><html><body><pre class="r"># The node label is assigned on genus level here; unlabeled nodes do not have a genus assigned.</pre></body></html></div>
<p>While the condition-specific microbial networks are not too informative, we can learn more from the global network computed on all samples, since in this case, we also capture species that increase or decrease in abundance across the two conditions (healthy vs IBD), thereby covarying or excluding each other. In order to know which taxa increase or decrease in IBD samples, we introduce a binary variable, named IBD status, that we derive from the group assignment vector and append to the abundance matrix. Note that the abundances with metadata appended should not be transformed, but since they already represent filtered relative abundances, no further transformation step is necessary. Please also note that we treat a binary as a numeric variable here, which means we interpret it as an “IBD status” quantity that takes two values, low and high. This approach is not permissible with the three original sample groups, since it imposes an order that is not naturally present in UC and CD samples (i.e. we cannot say that in UC samples, the “IBD status” is higher or lower than in CD samples). Interestingly, we see several instances of mutual exclusion between species belonging to the same genus, which we did not find in the condition-specific networks. It is possible that different representatives of the same genus do better in the healthy and the IBD samples, respectively, an outcome that would not have been visible without species-level resolution. In addition, we find that several species are inversely correlated to IBD status, including Coprococcus catus, Gordonibacter pamelaeae, Alistipes shahii and Roseburia hominis. These would deserve further exploration.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># we binarise the group assignment to append it to the abundances</span>
<span class="no">ibd.status</span><span class="kw">=</span><span class="no">ibd.groups</span>
<span class="no">ibd.status</span>[<span class="no">ibd.groups</span><span class="kw">==</span><span class="st">"Control"</span>]<span class="kw">=</span><span class="fl">0</span>
<span class="no">ibd.status</span>[<span class="no">ibd.groups</span><span class="kw">==</span><span class="st">"IBD"</span>]<span class="kw">=</span><span class="fl">1</span>
<span class="no">ibd_taxa_extended</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(<span class="no">ibd_taxa</span>[<span class="no">indices.prev</span>,],<span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="no">ibd.status</span>))
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">ibd_taxa_extended</span>)<span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">ibd_taxa</span>)[<span class="no">indices.prev</span>],<span class="st">"ibd.status"</span>) <span class="co"># append row name for IBD status</span>
<span class="no">ibd_lineages_extended</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(<span class="no">ibd_lineages</span>[<span class="no">indices.prev</span>,],<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"IBD_status"</span>,<span class="fl">7</span>)) <span class="co"># entry in lineage matrix needed to label node</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">ibd_lineages_extended</span>)<span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">ibd_lineages</span>)[<span class="no">indices.prev</span>],<span class="st">"ibd.status"</span>) <span class="co"># append row name for IBD status</span>
<span class="no">network</span><span class="kw">=</span><span class="fu"><a href="../reference/buildNetwork.html">buildNetwork</a></span>(<span class="no">ibd_taxa_extended</span>, <span class="kw">lineages</span><span class="kw">=</span><span class="no">ibd_lineages_extended</span>, <span class="kw">method</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"spearman"</span>,<span class="st">"bray"</span>), <span class="kw">repNum</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">nameLevel</span><span class="kw">=</span><span class="st">"species"</span>) <span class="co"># run CoNet on all samples</span></pre></body></html></div>
<pre><code>## [1] "Network construction with method(s): spearman, bray"
## [1] "Upper and lower threshold are ignored when more than one method is selected."
## [1] "Processing method spearman"
## [1] "Lower threshold for initial edge number (100): -0.328048640295398"
## [1] "Upper threshold for initial edge number (100): 0.517808720748735"
## [1] "Renormalisation is FALSE"
## [1] "P-values of correlations are computed with cor.test FALSE"
## [1] "Permutations and bootstraps are both computed FALSE"
## [1] "Processing method bray"
## [1] "Lower threshold for initial edge number (100): 0.685625933359747"
## [1] "Upper threshold for initial edge number (100): 0.997021090321542"
## [1] "Renormalisation is FALSE"
## [1] "P-values of correlations are computed with cor.test FALSE"
## [1] "Permutations and bootstraps are both computed FALSE"
## [1] "Associations computed for 110 taxa."
## [1] "Merging methods..."
## [1] "Network has 31 edges."</code></pre>
<p><img src="figure_seqgroup_tour/unnamed-chunk-18-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">network</span>,<span class="kw">main</span><span class="kw">=</span><span class="st">"Global network"</span>)</pre></body></html></div>
<p><img src="figure_seqgroup_tour/unnamed-chunk-18-2.png" width="576" style="display: block; margin: auto;"></p>
<p>Finally, we can look at the cross-correlation between species and function abundances using the barebones CoNet function directly. We take several shortcuts here to compute the cross-correlation network in a reasonable amount of time: first, we strongly filter the functions to keep only those that are present in most of the samples. Second, as above, we will not assess the significance of the correlations. The result is a bipartite network with two node sets, one for the species and one for the functions.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">prev</span><span class="kw">=</span><span class="fl">150</span> <span class="co"># minimum occurrence for functions</span>
<span class="no">ibd_functions_incidence</span><span class="kw">=</span><span class="no">ibd_functions</span>
<span class="no">ibd_functions_incidence</span>[<span class="no">ibd_functions_incidence</span><span class="kw">&gt;</span><span class="fl">0</span>]<span class="kw">=</span><span class="fl">1</span>
<span class="no">rowSums</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">ibd_functions_incidence</span>)
<span class="no">indices.functions.prev</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">rowSums</span><span class="kw">&gt;=</span><span class="no">prev</span>)
<span class="co"># CoNet expects metadata to have samples as rows, therefore functions have to be transposed</span>
<span class="no">bip.network</span><span class="kw">=</span><span class="fu"><a href="../reference/barebonesCoNet.html">barebonesCoNet</a></span>(<span class="kw">abundances</span><span class="kw">=</span><span class="no">ibd_taxa</span>[<span class="no">indices.prev</span>,],<span class="kw">metadata</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/transpose-methods.html">t</a></span>(<span class="no">ibd_functions</span>[<span class="no">indices.functions.prev</span>,]),<span class="kw">min.occ</span><span class="kw">=</span><span class="fl">0</span>,<span class="kw">methods</span><span class="kw">=</span><span class="st">"pearson"</span>,<span class="kw">T.up</span><span class="kw">=</span><span class="fl">0.7</span>,<span class="kw">T.down</span><span class="kw">=</span>-<span class="fl">0.7</span>)</pre></body></html></div>
<pre><code>## [1] "Network construction with method(s): pearson"
## [1] "Processing method pearson"
## [1] "Renormalisation is FALSE"
## [1] "P-values of correlations are computed with cor.test FALSE"
## [1] "Permutations and bootstraps are both computed FALSE"
## [1] "up and down"
## [1] "Associations computed for 109 taxa and 612 metadata."
## [1] "Network has 49 edges."</code></pre>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">bip.network</span>,<span class="kw">main</span><span class="kw">=</span><span class="st">"Global species-function network"</span>)</pre></body></html></div>
<p><img src="figure_seqgroup_tour/unnamed-chunk-19-1.png" width="960" style="display: block; margin: auto;"></p>
<p>Although the network is crudely built, the results are interesting. For instance, Akkermansia, which is a known mucin degrader, is associated to beta-N-acetylhexosaminidase. N-acetyl-D-glucosamine is a component of mucin. For Eubacterium rectale, three of its associated functions (with EC numbers 2.7.1.156, 2.7.7.62 and 3.1.3.73) are involved in vitamin B12 biosynthesis. Finally, Faecalibacteriun prausnitzii, a known butyrate producer, is associated to hydroxyl-butyryl-CoA dehydrogenase, which is involved in butyrate synthesis. Note that the abundances of the functions have been extracted from metagenomic and not metatranscriptomic data, which means that it is not clear whether enzymes encoded by these functional genes are also expressed.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Karoline Faust.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
